### PR TITLE
Binary Hash Join operator

### PR DESCRIPTION
**Pull request by srsuryadev**
### Summary
* Since the default join operator is a multiway join, wrote another flexible binary hash join operator which can be composed as a chain of join operators for the query.
* Renamed the old operator and join predicate members.
* Removed the ordering of table withing the operator and moved to the resolver.
* Unit tests for the binary hash join.

Main file in the PR :**src/operators/join/hash_join.cc**

### CODE REVIEW

File: src/utils/parallel_utils.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef HUSTLE_PARALLEL_UTILS_H
+#define HUSTLE_PARALLEL_UTILS_H
+
+namespace hustle {
+namespace utils {
+
+inline void ComputeBatchConfig(int &batch_size, int &num_batches,
```
Comment (GindaChen) [inline src/utils/parallel_utils.h:24]: Got it. I think hash aggregate will also use this

File: src/operators/join/hash_join.cc [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+          *filter->chunk(j)->data(),
+          arrow::compute::FilterOptions::NullSelectionBehavior::DROP);
+    }
+    hash_table_size = updated_length;
+  }
+  hash_table_->reserve(hash_table_size);
+
+  for (std::size_t i = 0; i < col->num_chunks(); i++) {
+    // Each task inserts one chunk into the hash table
+    // TODO: for now, we assume the join column is INT64 type.
```
Comment (GindaChen) [inline src/operators/join/hash_join.cc:124]: Got it - will attemp to generalize this in later PR.

File: src/operators/join/hash_join.cc [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+    left_indices_[i] = std::vector<uint32_t>(l_idxs, l_idxs + num_indices);
+    right_indices_[i] = std::vector<uint32_t>(r_idxs, r_idxs + num_indices);
+
+    free(l_idxs);
+    free(r_idxs);
+  }
+}
+
+void HashJoin::ProbeHashTable(
```
Comment (GindaChen) [inline src/operators/join/hash_join.cc:199]: Shall we later separate hash join into two phase (as two operators) - hash build + hash probe?

**Review by GindaChen — APPROVED** [review]
Overall looks good! LGTM!

File: src/utils/parallel_utils.h (hunk 1 above)
Comment (srsuryadev) [reply src/utils/parallel_utils.h:24]: Yes, other operators also use this for parallelism, will refactor them in the upcoming PRs

File: src/operators/join/hash_join.cc (hunk 2 above)
Comment (srsuryadev) [reply src/operators/join/hash_join.cc:124]: No issues, I am planning to work on providing support multiple types in Join in the coming PRs.

File: src/operators/join/hash_join.cc (hunk 3 above)
Comment (srsuryadev) [reply src/operators/join/hash_join.cc:199]: Sure, will do them in the coming work

**Comment by srsuryadev** [comment]
@GindaChen  Thank you for the review!

**Pull request merged**